### PR TITLE
Treenome Browser v1 integration

### PR DESCRIPTION
**Pull request by amkram**
Here's a PR where we can iron out the final details of the treenome browser.

A couple recent additions:
- added nt mutations (now the mutations displayed in the browser match those selected in the cog menu). Mutations also load progressively in "chunks" rather than all at once.
- added a band that appears at the selected node when zoomed in far enough, highlighting that node's genome

I think the last main thing to figure out is, as you mentioned, handling non-SARS-CoV-2 trees.

Right now it really only works for the cov2tree tree. The two main things needed to build the browser for an arbitrary tree are (1) the reference sequence and (2) coordinates of genes/CDS to compute nt positions of aa mutations. (2) is only necessary for aa mutations, so just displaying nt mutations could work without it.

I currently compute the reference through the fake "mutations" at the root node -- which I realize aren't guaranteed to be there (but I think they should be if a .gb file was used to make the tree)

(2) is hard-coded for SARS-CoV-2. But if we stored the CDS data from a .gb file, I could use those.

What do you think about storing the genbank CDS features (potentially also the fasta sequences) in the jsonl file, and only displaying the treenome toggle button if they are present?

### CODE REVIEW

File: taxonium_web_client/src/Taxonium.jsx
Comment (theosanderson) [inline taxonium_web_client/src/Taxonium.jsx:94]: Nit: I wonder if we should call this e.g. `gBrowserState` throughout (or `genomeBrowserState` but maybe that's too long). Atm it always makes me think of the web-browser since that's such a common thing.

File: taxonium_web_client/src/components/MutationHoverTip.jsx
Comment (theosanderson) [inline taxonium_web_client/src/components/MutationHoverTip.jsx:32]: I think I like this change to not have these colours in the tooltip. We can lose the comment once we're confident about that.

File: taxonium_web_client/src/components/SearchPanel.jsx
Comment (theosanderson) [inline taxonium_web_client/src/components/SearchPanel.jsx:168]: I think if we go with `Treenome browser` it would be good to have a `title` that explains what this does.

File: taxonium_web_client/src/Taxonium.jsx
Comment (amkram) [reply taxonium_web_client/src/Taxonium.jsx]: Good call. Switched to `genomeBrowser` throughout

File: taxonium_web_client/src/components/SearchPanel.jsx
Comment (amkram) [reply taxonium_web_client/src/components/SearchPanel.jsx:168]: We could just call it Genome Browser in the app to avoid confusion -- though I think we will end up using the name Treenome Browser in publication, so might be good to match. Let me know what you think. 

If we keep the name, maybe `Treenome Browser (?)` with the ? being either a link to a section in the docs or a tooltip with a brief description?

File: taxonium_web_client/src/components/MutationHoverTip.jsx
Comment (amkram) [reply taxonium_web_client/src/components/MutationHoverTip.jsx:32]: Yeah, It was pretty difficult to read when the colors were on the lighter side

File: taxonium_web_client/src/hooks/useBrowserLayerData.js
Comment (amkram) [reply taxonium_web_client/src/hooks/useBrowserLayerData.js]: done

File: taxonium_web_client/src/components/SearchPanel.jsx
Comment (theosanderson) [reply taxonium_web_client/src/components/SearchPanel.jsx:168]: Yeah that sounds great to me! I think tooltip would be great, but we can do custom tooltip that includes a link to docs.

**Comment by theosanderson** [comment]
Thanks for the PR! – I launched straight into nits before posting this appreciative comment :) Looking great. More to come (not too much hopefully!) from me.

File: taxonium_web_client/src/components/MutationHoverTip.jsx
Comment (theosanderson) [reply taxonium_web_client/src/components/MutationHoverTip.jsx:32]: Nice

File: taxonium_web_client/src/Taxonium.jsx
Comment (theosanderson) [reply taxonium_web_client/src/Taxonium.jsx]: Thanks!

Comment (amkram) [reply taxonium_web_client/src/Taxonium.jsx]: changed this again -- to `treenome` throughout for consistency (and shorter names)

File: taxonium_web_client/src/hooks/useBrowserLayers.js
Comment (amkram) [reply taxonium_web_client/src/hooks/useBrowserLayers.js]: At one point early on I decided on polygon for some reason... Tried again with line layers, works great and no obvious reason not to. Makes more sense to use lines, and seems (?) like there might be some speedup on rendering.

File: taxonium_web_client/src/Taxonium.jsx
Comment (theosanderson) [reply taxonium_web_client/src/Taxonium.jsx]: This is great - genome browser also had an implication it might apply just to the top panel

**Comment by theosanderson** [comment]
Re: linting, please run `eslint .` or `yarn eslint .` in the `taxonium_web_client` dir. It highlights some issues it would be great to address, in part just to avoid them spamming us with lint issues in future. I had a handful of lint failures before, so don't worry about those (and todo on me to fix). If you use VS-code, the esLint extension is a good way to see these during dev. In case you have issues running this, the current output is: https://pastebin.com/BjdnWU5B .

**Comment by amkram** [comment]
> Re: linting, please run eslint . or yarn eslint ....

Thanks for the tip about the VSCode extension, super useful. I believe I fixed most of the linting warnings caused by the treenome code. There is a dependency array issue (`JBrowsePanel.jsx:230`) that causes a weird feedback loop if I "fix" it, so I will see if I can restructure that code somehow (side note, I think that `eslint .` by default only picks up the `.js` files rather than `.jsx` -- I'm seeing a few additional warnings from `.jsx` files in the console output of `yarn start`) 

Also: do you also see this error `Failed to parse source map from '/Users/alex/Science/taxonium/taxonium_web_client/node_modules/readable-web-to-node-stream/lib/index.js.map' file: Error: ENOENT: no such file or directory, open '/Users/alex/Science/taxonium/taxonium_web_client/node_modules/readable-web-to-node-stream/lib/index.js.map'`

when running the app with `yarn start` ? I have been seeing it for a while, things still run fine but not sure if it was caused by treenome stuff or not...

**Comment by theosanderson** [comment]
> I believe I fixed most of the linting warnings caused by the treenome code.

Thanks!

> There is a dependency array issue (`JBrowsePanel.jsx:230`) that causes a weird feedback loop if I "fix" it, so I will see if I can restructure that code somehow 

Yeah I've encountered things like that. Feel free to just put in a comment that disables eslint for that line, which I have doen at least in the past. From chatting with an experienced React dev, apparently this is sort of an abuse of dependency arrays and we should add extra state variables instead to track whether we should run again, but.. it's fine by me :)


> (side note, I think that `eslint .` by default only picks up the `.js` files rather than `.jsx` -- I'm seeing a few additional warnings from `.jsx` files in the console output of `yarn start`)

Ah, yeah I was actually relying on the VS code stuff. There is probably stuff I need to do to reconfigure the environment properly.

> Also: do you also see this error `Failed to parse source map from '/Users/alex/Science/taxonium/taxonium_web_client/node_modules/readable-web-to-node-stream/lib/index.js.map' file: Error: ENOENT: no such file or directory, open '/Users/alex/Science/taxonium/taxonium_web_client/node_modules/readable-web-to-node-stream/lib/index.js.map'`
> 
> when running the app with `yarn start` ? I have been seeing it for a while, things still run fine but not sure if it was caused by treenome stuff or not...

Yeah, don't worry I don't think it's caused by Treenome, it's https://github.com/facebook/create-react-app/discussions/11767 , presumably due to an issue in that package. It has sent me down a bunch of rabbit holes when unrelated things are broken. I will consider changing to disable the warning because as you say, it doesn't matter.

File: taxonium_web_client/src/components/SearchPanel.jsx
Comment (theosanderson) [reply taxonium_web_client/src/components/SearchPanel.jsx]: Nice!

Comment (theosanderson) [reply taxonium_web_client/src/components/SearchPanel.jsx]: Can link to https://docs.taxonium.org/en/latest/treenome.html and add a placeholder with even a couple of sentences  in `/docs/treenome.md` if you are happy to use the Taxonium docs. (Using those docs would be my minor preference, but if you prefer an external site that's fine by me).

**Comment by theosanderson** [comment]
Let me know if the new release of taxoniumtools doesn't give you what you need

**Comment by amkram** [comment]
Latest commits should make Treenome work on any tree made with `usher_to_taxonium` with a genbank file. I could also make it work with nextstrain imports (by e.g. creating fake root mutations, which I'm still relying on for .gb trees) but maybe better to wait on that until the switch to encoding the sequence directly in the json?

File: taxonium_web_client/src/components/SearchPanel.jsx
Comment (theosanderson) [reply taxonium_web_client/src/components/SearchPanel.jsx:283]: @amkram Pls can we try to do this before merging

Comment (amkram) [reply taxonium_web_client/src/components/SearchPanel.jsx:283]: Added some docs to `treenome.md`

Comment (amkram) [reply taxonium_web_client/src/components/SearchPanel.jsx:283]: I also added a short tutorial video which might be useful, as I think navigating treenome browser is slightly less intuitive than base Taxonium

**Comment by theosanderson** [comment]
Looking really great. I'll try to reserve almost all further comments for https://github.com/theosanderson/taxonium/issues/303 after merging.

I notice the sidebar width got smaller compared to master, even in the absence of the treenome browser. It would be great to revert that unless you disagree. On e.g. a Macbook pro the Search wraps now which isn't completely ideal. 
image

That's the last thing I can think of at the moment than needs to happen before merging. I'll post some stuff in the enhancements now.

**Comment by theosanderson** [comment]
Something else, not sure how hard to fix: when you hit the horizontal zoom buttons on standard Cov2Tree it zooms on the centre of the UI. This doesn't seem to happen for Treenome -- probably it is zooming on where the center would be if the treenome wasn't visible

**Comment by theosanderson** [comment]
One more: a kind of cool thing about Taxonium is that it vaguely works on mobile, and I think that makes a difference to initial user adoption so I'd like to keep that going. The treenome browser bit actually works really well, to my surprise. But I think the sidebar changes have resulted in not using all the screen real estate, especially when the "sidebar" is closed.

**Comment by amkram** [comment]
Re: sidebar, last commit should revert that back closer to Taxonium master's. I also made the sidebar toggle up and down in mobile mode (determined by the screen width < 600 check which you seem to have used elsewhere)

**Comment by theosanderson** [comment]
> Re: sidebar, last commit should revert that back closer to Taxonium master's. 

Nice -- if you want to shrink it when treenome is enabled that's totally cool with me - it was just the change without the box ticked that didn't suit me

**Comment by theosanderson** [comment]
The mobile down arrow is really cool. Nit: On my Pixel 6 it prevents you ever opening the bar again -- we should try to get the reopen button to persist somewhere, possibly just sitting over the deck in the bottom left corner. But OK to leave that to me to deal with at some point after the merge.

File: taxonium_web_client/src/components/SearchPanel.jsx
Comment (theosanderson) [reply taxonium_web_client/src/components/SearchPanel.jsx:283]: (Awesome stuff! - I was thinking I need videos too even before this and now even more convinced)

**Comment by amkram** [comment]
I bumped up the down arrow a bit. It looks good on my Pixel 5 at least, but possibly not the case with all devices.

> Ed: just playing and it probably does need to be narrower when Treenome is open (or else some other JBrowse edit) otherwise you don't get the full JBrowse zoom controls on a Macbook pro resolution.

I had forgotten that was why I initially made the sidebar narrower! Now toggling Treenome shrinks the sidebar a bit while it remains full-width with Treenome disabled.

**Comment by theosanderson** [comment]
> Now toggling Treenome shrinks the sidebar a bit while it remains full-width with Treenome disabled.

Very nice - apologies for making things complicated!

**Comment by theosanderson** [comment]
By the way we don't have to block on the zoom issue - let me know if you want to merge nowish

**Comment by amkram** [comment]
I think I have an idea how to fix the zoom issue -- will try a couple things over the weekend and reconvene soon

**Comment by amkram** [comment]
Ok, latest commit should fix that horizontal zoom issue!

**Review by theosanderson — APPROVED** [review]
LGTM! Thanks for all the hard work!

**Comment by theosanderson** [comment]
Oh yeah I remember this thing about the build failure to Docker. I looked into that a bit before: it's because of a canvas dependency from JBrowse. Will have a think about that..

**Comment by theosanderson** [comment]
> Oh yeah I remember this thing about the build failure to Docker. I looked into that a bit before: it's because of a canvas dependency from JBrowse. Will have a think about that..

What might work is to: set https://github.com/theosanderson/taxonium/blob/master/Dockerfile.frontend#L1 to `node:18.0.0-alpine` which should be v102.  That should then let us use [canvas-v2.9.3-node-v102-linux-glibc-x64.tar.gz](https://github.com/Automattic/node-canvas/releases/download/v2.9.3/canvas-v2.9.3-node-v102-linux-glibc-x64.tar.gz) if we pin canvas to 1.2.93

**Comment by amkram** [comment]
It [seems](https://github.com/node-gfx/node-canvas-prebuilt/issues/77) like it is due to the fact that alpine uses `musl` instead of `glibc`, and there is no `musl` build for canvas.  I think using e.g. a [bullseye image](https://hub.docker.com/_/node) instead of alpine would probably solve that, if you are open to the switch? Otherwise I think we can maybe build canvas from source

**Comment by theosanderson** [comment]
Ah good catch – definitely fine with that!

**Comment by theosanderson** [comment]
I'm not depending on the Docker image at the moment -- that was in use on a version within GISAID but that project seems to have stalled -- so feel free to merge this without fixing that and I can try to figure it out in due course

**Comment by amkram** [comment]
Looks like the image is building okay now.

Another issue was that canvas [doesn't have builds](https://github.com/Automattic/node-canvas/issues/2025) for Node 18 yet. As far as I can tell, in the [releases](https://github.com/Automattic/node-canvas/releases/), `node-v102` corresponds to Node 16, and `node-v108` would be Node 18 (if it were present). I downgraded the docker image to Node 16 which builds successfully.

Let me know if you think downgrading to 16 is an issue.

I don't know enough about node/docker to know why this was only an issue in the Docker build...

**Comment by theosanderson** [comment]
>`node-v102` corresponds to Node 16, and `node-v108` would be Node 18 (if it were present). I downgraded the docker image to Node 16 which builds successfully.
Apologies that I got this wrong!
> Let me know if you think downgrading to 16 is an issue.
Hmm interesting. It definitely would be an issue for the backend, which uses the fetch API (which is normally in the browser, not node, but was added experimentally in node 17 and properly in node 18). Somehow I convinced myself it was needed generally but I think you're right that it's not.
> I don't know enough about node/docker to know why this was only an issue in the Docker build...
Yeah me too :(

**Comment by theosanderson** [comment]
> BTW, should the integration tests be failing? I think it's due to a permissions issue (I had removed my write access so I wouldn't accidentally push to main again 😬 )

Yeah this is permissions but not to worry -- I think it will be fine, we can definitely merge as is. Are you happy for me to merge now?

**Comment by amkram** [comment]
Awesome. I am ready if you are!

**Comment by theosanderson** [comment]
OK, let's see how this goes! Thanks for the amazing feature and working through all these points.

**Comment by theosanderson** [comment]
🎉 🎉 everything seems to be working, tests passing, etc.

**Comment by amkram** [comment]
Yay! Thanks for all of your help and patience!

**Pull request merged**
